MSI InventorySales to take into account reservations when determining backorder quantity

gwharton opened the pull request:
### Description (*)
Add new GetBackOrderQty class to determine the number of back ordered items for new order, given current stock levels and reservations.

Change BackOrderNotifyCustomerCondition and BackOrderQuantityCondition classes to use new GetBackOrderQty function.

Move code from CheckQuoteItemQtyPlugin into own StockStateProvider Class. CheckQuoteItemQtyPlugin now calls this class instead of implementing own functionality. New StockStateProvider class correctly sets ItemBackorders data value fixing issue #2218.

New plugin ProviderCheckQuoteItemQtyPlugin calls new StockStateProvider Class.

Added new Integration tests for backorders

Added new website_id extension attribute to StockItemInterface to allow website ID to be accessed by backorder functions. This is missing on one of the legacy calls that are now intercepted.

### Fixed Issues (if relevant)
1. magento/inventory#2218 : Because of inventory_sales module (MSI) : Backorder Item Status is not showing if we placed order with backorder item in magento2.3.x
2. magento/inventory#2454 : Reservations not being taken into account properly on backorder warning/saleable qty/new orders.

### Manual testing scenarios (*)
1. See issue 2218, 2254

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds are green)

magento-cicd2 commented:
[![CLA assistant check](https://cla.dev.magento.com/pull/badge/signed)](https://cla.dev.magento.com/magento/inventory?pullRequest=2452) <br/>All committers have signed the CLA.

gwharton commented:
This is an order for 4 items, 1 of which is backordered.

![image](https://user-images.githubusercontent.com/30697781/62628005-69689900-b922-11e9-94df-fe52e1101106.png)

And an order for 1 item with 1 on backorder.

![image](https://user-images.githubusercontent.com/30697781/62628055-7dac9600-b922-11e9-89be-653ac929a990.png)

So in my limited test case, with my settings, it seems to work fine :)

gwharton commented:
And an order for 1 item with 1 on backorder.

![image](https://user-images.githubusercontent.com/30697781/62627933-40480880-b922-11e9-8360-66ce48760cc1.png)

ishakhsuvarov commented on app/code/Magento/InventorySales/Model/IsProductSalableCondition/BackOrderNotifyCustomerCondition.php:
```diff
@@ -74,15 +74,20 @@ public function execute(string $sku, int $stockId, float $requestedQty): Product
                 ];
-                return $this->productSalableResultFactory->create(['errors' => $errors]);
+                $result = $this->productSalableResultFactory->create(['errors' => $errors]);
+                $result->setBackOrderQty($backOrderQty);
```
While existing implementation of `ProductSalableResultInterface` is built on top of `DataObject` and allows to set arbitrary data into it, it's still not the best practice to use methods not defined in the contracts.

It's possible to overcome it in a more clean way using an extension attribute.

gwharton replied:
Implemented as requested in last commit.

ishakhsuvarov commented on app/code/Magento/InventorySales/Model/IsProductSalableCondition/BackOrderNotifyCustomerCondition.php:
```diff
@@ -74,15 +74,20 @@ public function execute(string $sku, int $stockId, float $requestedQty): Product
+                        [
                             'code' => 'back_order-not-enough',
                             'message' => __(
                                 'We don\'t have as many quantity as you requested, '
```
It's clear, that you were not an author of this code, but still, it would make sense to clean up this concatenated string which would not work, while we are here :)

gwharton replied:
Can you explain in a little more detail what it is you are after here.

ishakhsuvarov replied:
Please disregard my previous comment. It should be fine as is and I wasn't attentive enough.

ishakhsuvarov commented:
@gwharton Thank you for this Pull Request. This solution seems to work out in a better way then the previous fix. 

Currently I don't have any major comments, apart from the fact that some non-contract methods are called.

ishakhsuvarov commented:
Thanks @gwharton 

Looks good to me at this point!

ishakhsuvarov commented:
Some sort of an automated test would be really helpful in this case, however I'm also going to request a manual testing.

smoskaluk commented:
Hi @gwharton!
After manual testing I see that issue fixed only for products assigned to default source, but not on custom sources. Please recheck you PR.

gwharton commented:
OK, I'm about to commit some changes to this PR. Firstly, I'm going to explain what I've done as an initial look at the commit, it would appear that alot has changed, and it has (kindof).

Take a look at this piece of code. It is in the legacy CatalogInventory Quote Item initialization routines that get called when quotes are created. See that it calls `$this->stockState->checkQuoteItemQty` which is intercepted by the InventorySales module. All good here. Now look at the lines after. It calls `$this->stockStateProvider->checkQuoteItemQty` to get the amount of backorders. StockStateProvider is NOT intercepted and is processed with legacy code. This is the reason why backorder quantities are not correct (#2454) as the legacy code called by StockStateProvider is using product quantity, not saleable quantity. The correct backorder quantity returned by the InventorySales `stockState->checkQuoteItemQty` call is being overwritten with the legacy backorder quantity, which isn't the same thing.

https://github.com/magento/inventory/blob/4eed86a0b93e905e41744f6d77de1943c8f22731/app/code/Magento/CatalogInventory/Model/Quote/Item/QuantityValidator/Initializer/StockItem.php#L112-L124

So, the first change is that I moved the code that was in the StockState interceptor into its own Model class \Magento\InventorySales\Model\Stock\StockStateProvider, and then updated the stockState interceptor to call the new StockStateProvider class, and then added a new Interceptor so both legacy calls to `$this->stockState->checkQuoteItemQty` and `$this->stockStateProvider->checkQuoteItemQty` are both routed through the new StockStateProvider class.

OK, Problem Number 1. As we can't change any code in legacy CatalogInventory classes, were stuck with the legacy parameters that are passed to the call `$this->stockStateProvider->checkQuoteItemQty`. This does not include the website_id, and it is required by the new stockStateProvider class (website_id is passed by the `$this->stockState->checkQuoteItemQty` call). The only way I could see of achieving this without altering legacy code is to create a beforeInitialize plugin for the StockItem, which sets an extension attribute for the website ID. This means that the StockItem object that is passed to the `$this->stockStateProvider->checkQuoteItemQty` call now contains the website_id and can be accessed by the new StockStateProvider class.

Problem Number 2. My initial implementation of the fix added items to the BackOrderNotifyCustomerCondition class. Like the name suggests, this would only process the changes that I made under the circumstances that Notify was enabled, which is not in all cases where backorders are enabled. Therefore, I created a new class which is responsible purely to return the number of items backordered, which works in either notify and non-notify modes. This is the new BackOrderQuantityCondition class. The new StockStateProvider class checks for backorders completely independently of whether a backorder message is to be displayed.

I have added additional test assertions to the Magento\InventorySales\Test\Integration\Order tests so that it checks the backorder quantities on the order object, and verifies that the numbers are correct.

I have added additional tests to the Magento\InventorySales\Test\Integration\SalesQuoteItem tests which confirms that when items are added to the cart which would result in a backorder, that the correct backorder quantity and backorder message is added to the quote item.

 Commit coming shortly.

gwharton commented:
Seems there are functional test failures related to cart price rules. I'll investigate and commit update.

smoskaluk commented:
Hi @gwharton! Thank you for updating your PR, I see that problem with status of backordered order items is fixed. Only one thing I found out is not correct - the message when you create order via admin. When I adding product to the order - I got two messages - first is correct, and second unnecessary and wrong message with doubled qty of adding item:
 
![BFCE7852-9C64-4B5E-BD41-A6034BADCD18](https://user-images.githubusercontent.com/27818516/63276768-b77b8600-c2ac-11e9-9582-9f45b72c8d1f.jpg)

gwharton commented:
OK, leave that with me.

gwharton commented:
OK, that's an existing bug in the Magento\Sales module.

https://github.com/magento/inventory/blob/4eed86a0b93e905e41744f6d77de1943c8f22731/app/code/Magento/Sales/Block/Adminhtml/Order/Create/Items/Grid.php#L130-L143

That code snippet is run when the products grid is initialised on the admin pages when adding products to an order. Notice the little snippet on Line 137. Looks like someone wanted to dispatch an event, and thought the best way of doing it is to set the quantity of the item on the quote to the same quantity, afterall thats not changing anything........ right!........wrong!.

Updating the quantity on the quote actually causes the quote to be revalidated through the check functions, including checking for backorders etc. However what seems to be happening by doing that, is that it effectively validates the quote for the new qty without first removing the existing quantity. Presumably it thinks you require qty x, which is on the existing quote, and now you're asking for another qty x. Ofcourse this triggers off the backorder check for the wrong quantity, hence the extra message. The grid shouldn't be triggering the update of the quote at all.

The grid is updated with the correct backorder message from the initial quotation which is correct, and then it is updated with an erroneous backorder message after it updated the quote on Line 137.

It is present in mainline 2.3-develop. Issue raised.

https://github.com/magento/magento2/issues/24191

smoskaluk commented:
Hi @gwharton! How I see, issue described in https://github.com/magento/inventory/issues/2454 still reproducing (backorders still not take in account reservations of previously placed orders, without source deduction through the shipment)

gwharton commented:
Blinded by the other issues, forgot about checking that. Will look into it today.

gwharton commented:
Hold on further progression on this. Will be checking in integration test for above commit shortly

gwharton commented:
@magento run all tests

smoskaluk commented:
Hi @gwharton! I have tested your PR with latest commit and found next things:
1. Wrong calculation of backordered items in message in shopping cart on frontend when there are previously placed reservations for product
![backorders1](https://user-images.githubusercontent.com/27818516/63514245-d66a5a00-c4f0-11e9-9249-a4022adc9d1e.png)

2. The same wrong calculation of backordered items in message when placing order from admin with product on custom stock/source
![backadmin2](https://user-images.githubusercontent.com/27818516/63514515-6d371680-c4f1-11e9-9ae9-37dc0ef68635.png)

3. Wrong unnecessary message - "Product out of stock" - when placing order from admin with product on default stock/source
![backadmin1](https://user-images.githubusercontent.com/27818516/63514643-b6876600-c4f1-11e9-9aaf-056a321262f9.png)

4. Also, as you may see on screenshots, messages for default stock and custom stock are not similar
5. Order items if backordered from custom stock - in order have status - partial
![partial](https://user-images.githubusercontent.com/27818516/63515204-ef740a80-c4f2-11e9-8646-6890f768234b.png)

gwharton commented:
Crikey. Right. Leave them with me. Im away until late next week so will take a look then.

I know i changed it to use salable qty, which i guess is wrong when there is min qty set less than zero. I guess it should be using salable qty plus min. So if you have min set to -20 with 5 in stock salable qty would be 25 and you order 6, the backorder amount would be 1. Not sure that explains all of your findings so will analyse them all.

Ill be a bit more thorough and get it right this time. These things always seem to turn out trickier than initially expected. I was hoping this was gonna be a one liner!

gwharton commented:
I suspect some of what you are seeing is related to mainline issue https://github.com/magento/magento2/issues/24203 where adding additional quantities to existing quote items is broken for backorder quantity calculations. Will report back later.

gwharton commented:
I still have more work to do on this, so hold any future testing.

gwharton commented:
I still have more work to do on this, so hold any future testing.

gwharton commented:
I still have more work to do on this, so hold any future testing.

gwharton commented:
Known Issue 1 : When adding products to order under Adminhtml order creation page, Erroneous backorder messages are displayed. See Mainline Issue https://github.com/magento/magento2/issues/24191

Known Issue 2 : If you set Product to 10 in stock, Out of stock threshold to -10, Allow backorders below 0 with notify. Add 20 items to your cart. You should receive warning that 10 is being backorderd. If you change quantity in cart to 21 you will get a popup with both a warning that 10 is being backordered and an error stating qty not available. If you remove all items from cart and re-add 21 items, you get the correct error saying qty not available. See Mainline Issue https://github.com/magento/magento2/issues/24366

Known Issue 3 : When adding products to order under Adminhtml, if you order more stock than is available, e.g 10 in stock, Out of stock Threshold -10, and you add 30 items, the page allows you to add all 30 items to quote, and prints a warning that the "requested quantity is not available". That should be an error and it should disallow it. See Mainline Issue https://github.com/magento/magento2/issues/24365

gwharton commented:
> Hi @gwharton! I have tested your PR with latest commit and found next things:
> 
> 1. Wrong calculation of backordered items in message in shopping cart on frontend when there are previously placed reservations for product
>    ![backorders1](https://user-images.githubusercontent.com/27818516/63514245-d66a5a00-c4f0-11e9-9249-a4022adc9d1e.png)

Yes. I located where this was happening. The getBackOrderQuantity function was actually calculating the total number of items to be backordered, which includes the amount in the current quote, and also any reservations that exist which also contained backordered items. I have fixed this so the number of items backordered will only include the quantity requested on the quote.

> 2. The same wrong calculation of backordered items in message when placing order from admin with product on custom stock/source
>    ![backadmin2](https://user-images.githubusercontent.com/27818516/63514515-6d371680-c4f1-11e9-9ae9-37dc0ef68635.png)

Fixed as per Number 1 above.

> 3. Wrong unnecessary message - "Product out of stock" - when placing order from admin with product on default stock/source
>    ![backadmin1](https://user-images.githubusercontent.com/27818516/63514643-b6876600-c4f1-11e9-9aaf-056a321262f9.png)

I think this is related to Known Issue 2 above. I'm working on it.

> 4. Also, as you may see on screenshots, messages for default stock and custom stock are not similar

Not sure what's going on here. The message "We don't have as many quantity........" message appears when there is a backorder amount, but it does not breach the minqty (if less than zero). The "Product out of stock" message appears when the amount causes a breach of minqty, and is therefore not allowed.

> 5. Order items if backordered from custom stock - in order have status - partial
>    ![partial](https://user-images.githubusercontent.com/27818516/63515204-ef740a80-c4f2-11e9-8646-6890f768234b.png)

I've tested this but can't reproduce. If you add for example 4 items, but only 2 would be backordered, this would result in "Partial". Whereas if you add 4 items and all 4 are backordered, it would result in "Backordered"

gwharton commented:
Well, i think we're closer now. I will be doing more extensive testing on this tomorrow, but feel free to step in and do some testing aswell. There are so many variables, its not an easy one to ensure you are testing all variants.

There are 3 issues that still exist, however all are issues existing in Mainline Magento. All have issues raised. See previous post for details.

ishakhsuvarov commented on app/code/Magento/InventorySales/Model/GetBackOrderQty.php:
```diff
@@ -0,0 +1,61 @@
+use Magento\InventorySalesApi\Api\GetProductSalableQtyInterface;
+
+/**
+ * @inheritdoc
```
There is no class to inherit from. Description of class purpose would be really helpful here :)

gwharton replied:
Understood. Cut and paste strikes again. :) I will add necessary comments.

ishakhsuvarov commented on app/code/Magento/InventorySales/Model/GetBackOrderQty.php:
```diff
@@ -0,0 +1,61 @@
+            $salableQty = $this->getProductSalableQty->execute($sku, $stockId);
+            $minqty = $stockItemConfiguration->getMinQty();
+            $backOrderQty = $requestedQty - $salableQty - $minqty;
+            if ($requestedQty < $backOrderQty) {
```
We tend to stick to `bccomp` when comparing floats, just to make sure nothing strange happens in computation.

gwharton replied:
Sure, no problem. I see that elsewhere and will change.

ishakhsuvarov commented on app/code/Magento/InventorySales/Model/GetBackOrderQty.php:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    /**
+     * @inheritdoc
```
same as above

ishakhsuvarov commented on app/code/Magento/InventorySales/Model/IsProductSalableCondition/IsSalableQuantityCondition.php:
```diff
@@ -82,11 +72,7 @@ public function execute(string $sku, int $stockId): bool
-            $this->getReservationsQuantity->execute($sku, $stockId);
         return (bool)$stockItemData[GetStockItemDataInterface::IS_SALABLE] &&
-            $qtyWithReservation > $stockItemConfiguration->getMinQty();
+            $this->getProductSalableQty->execute($sku, $stockId) > 0;
```
`GetProductSalableQty` service seems to perform more checks then there were in this service

gwharton replied:
So, I thought hard about this. 

There seemed to be two strains of code developing in the module that did the exact same thing. We have the notion of SalableWithReservations and we also have SalableQuantity. Looking at both of these strains wherever they occurred in the codebase they both essentially did exactly the same thing to achieve the same purpose (work out how many of the item were available for sale (not how many in stock)).

I made these changes to try and keep the core logic of determining the actual quantity available for sale in a single place, i.e in the getProductSalableQty class as it currently exists, as the above code is using its own method to work out salableQuantity (directly using stock levels and reservations), and not calling the getSalableQuantity class to get that info. It made sense to remove the direct accessing of reservations from this function, and just make a call to check is salableQuantity is greater than 0.

I also changed the name of this class from IsSalableWithReservationsCondition to IsSalableQuantityCondition as a way of trying to force the use of SalableQuantity as the concept and notSalableWithReservations. (even though they are identical).

The user see's it as Salable Quantity, the developer should see it as SalableQuantity and not SalableWithReservations in my opinion.

Yes the getProductSalableQty function performs additional checks above what was originally in this class, but they are all relevant checks.

I realise this is a fairly substantial change, and would be pleased if this sparked further discussion within the team. This is just my opinion of some changes that I felt should be proposed, so there they are.

ishakhsuvarov commented on app/code/Magento/InventorySales/Model/IsProductSalableForRequestedQtyCondition/IsSalableQuantityCondition.php:
```diff
@@ -7,35 +7,23 @@
  * @inheritdoc
  */
-class IsSalableWithReservationsCondition implements IsProductSalableForRequestedQtyInterface
+class IsSalableQuantityCondition implements IsProductSalableForRequestedQtyInterface
```
This seems to be a massive change, how would we calculate reservations now?

gwharton replied:
The same discussion as above really. An attempt to standardise on the use of the SalableQuantity function to determine the quantity of stock for sale, instead of calculating it directly using stock levels and reservations (thats the job of the salableQuantity function)

If you work through the logic of the changes below and compare with whats going on inside getSalableQty its the same thing.

You can still access reservations directly, but if all that you want to know is how many can I sell (i.e the salable quantity) then use the salableQuantity functions (which takes into account current reservations)

Again, open for discussion. It just feels cleaner, more correct, less duplication of logic by making everything use getSalableQty instead of working directly with stock levels and reservations.

gwharton replied:
Happy to take the chat over to slack #msi if requred

gwharton commented:
Once all the issues are resolved, and if the decision is to accept this PR, give me a nod and I will git squash all the commits into a single commit to keep the commit history clean prior to merge.

gwharton commented:
Apologies if this whole PR has become too massive to review. I initially had intended it to be a one liner, but there are several issues resolved here that were uncovered along the way that have been fixed, along with some restructuring (with hindsight, I should have kept the restructuring separate from the bugfixing).

Let me know if you would prefer me to look at the possibility of splitting out the changes into individual PRs to make processing/reviewing easier.

gwharton commented:
I have reverted the "nice to have's" in this PR as I realise they may have been contentious renaming core classes. So this PR is now just the fixes required in order to make Backorders calculate corrcetly when there are active reservations, and also the fix to show backorder status in the admin GUI.

gwharton commented:
I'm not too sure what the way forward on the Magento Health Index failures. I added Magento/Quote as a dependency in the composer.json which results in this failure. I see other PR's having similar failures. Any clues?

sidolov commented:
Hi @gwharton , due to changes in the repository structure (Magento core code is extracted from the MSI code base) we have changed the base branch to `1.1-develop-legacy`. We will keep this branch for a while to make a smooth transition, in the meanwhile, please, apply changes from the current PR to the new code base structure in `1.1-develop` branch. 

[Installation Guide](https://github.com/magento/inventory/wiki/Metapackage-Installation-Guide) may be helpful for local project development. 

Let me know if you need help or have questions!
Thank you for your collaboration!

Commit:
Introduced new GetBackOrderQty class for BackOrder condition checks. Improved class documentation. Introduced StockStateProvider class and call this from ProviderCheckQuoteItemQtyPlugin and StockItemPlugin. Added various integration tests for backorders.

gwharton committed:
Merge branch '1.1-develop' into 2.3-develop

Commit:
Check for isManagedStock on backorder calculations in all cases.

Commit:
Modified Unit Tests to use getBackOrderQty classes instead of calculating
backorders directly.

gwharton committed:
Merge branch '1.1-develop' into 2.3-develop

gwharton commented:
Not sure whats going on with the integration tests. Ive rerun a number of times and still fails.

sidolov committed:
Merge branch '1.1-develop' into 2.3-develop

irenelagno committed:
MC-31098: Add support of Elasticsearch7 to codebase

Commit:
magento/inventory#2452: Fixed fixtures paths

Commit:
Merge remote-tracking branch 'origin/1.1-develop' into gwh-2.3-develop

igrybkov committed:
Fix exception when a product is not assigned to all stocks

nmalevanec committed:
MSI-2583[BackPort]: Unchecked Notify Quantity "Use Default" Not Saved after Product Save.

nmalevanec committed:
MSI-2583[BackPort]: Unchecked Notify Quantity "Use Default" Not Saved after Product Save.

nmalevanec committed:
Fix exception when a product is not assigned to all stocks. Cover with integration test.

nmalevanec committed:
Fix exception when a product is not assigned to all stocks. Fix test.

sidolov committed:
Merge pull request #2892 from nmalevanec/MSI-2583(BackPort)

MSI-2583[BackPort]: Unchecked Notify Quantity "Use Default" Not Saved after Product Save.

sidolov committed:
Fixed fixtures path

sidolov committed:
Merge pull request #2894 from magento-trigger/elastic_spike

MC-31098: Add support of Elasticsearch7 to codebase

nmalevanec committed:
Fix exception when a product is not assigned to all stocks. Fix test.

sidolov committed:
Merge branch '1.1-develop' into 1.1-product-not-on-all-sources

sidolov committed:
Merge pull request #2878 from igrybkov/1.1-product-not-on-all-sources

Fix exception when a product is not assigned to all stocks during IsOrderSourceManageable check

nmalevanec committed:
MSI-2904: Fatal error in `inventory:reservation:list-inconsistencies`because of wrong sku parameter type. [Backport]

sidolov committed:
Merge pull request #2912 from nmalevanec/MSI-2904(Backport)

MSI-2904: Fatal error in `inventory:reservation:list-inconsistencies` because of wrong sku parameter type. [Backport]

gwharton committed:
Updated integration test fixture files in accordance with PR #2931

gwharton committed:
Updated integration test fixture files in accordance with PR #2931

gwharton committed:
Updated integration test fixture files in accordance with PR #2931

gwharton committed:
Updated integration test fixture files in accordance with PR #2931

arnobsh committed:
MC-32285: Cannot assign more than 20 sources to product

- Updating status and quantity from the requested data
- MFTF test coverage

zakdma committed:
MC-33237: [2.3.6] Jenkins - tests permanently fail because of testing infrastructure issues

sidolov committed:
Merge pull request #2948 from magento-tsg/MC-33237

[TSG] MC-33237: [2.3] Jenkins - tests permanently fail because of testing infrastructure issues

bubasuma committed:
MC-33464: Inventory:reservation:list-inconsistencies uses wrong connection with split databases

- Fix connection name for queries to sales tables

bubasuma committed:
MC-33464: Inventory:reservation:list-inconsistencies uses wrong connection with split databases

- Remove direct usage of resource connection and replace it with appropriate resource model
- Fix remove JOIN queries between tables from different area

bubasuma committed:
MC-33464: Inventory:reservation:list-inconsistencies uses wrong connection with split databases

- Remove StoreWebsiteResolver service

Commit:
MC-32285: Cannot assign more than 20 sources to product

Commit:
Merge pull request #2951 from arnobsh/MC-32285

MC-32285: Cannot assign more than 20 sources to product

bubasuma committed:
Merge remote-tracking branch 'origin/1.1-develop' into MC-33464

Commit:
Merge pull request #2977 from magento-mpi/MC-33464

inventory:reservation:list-inconsistencies command fails with split database config

maghamed commented:
hey @Graham Wharton, sorry for making you wait for my answer. 
I looked though the codebase briefly and to issues, you referenced in your PR.

Like this one - https://github.com/magento/inventory/issues/2454
Do we still have this issue reproducible, btw? // cc @Stanislav Idolov 
Is this bug only reproducible with backorder enabled? At least from the top of my head I can’t find a quick reason how you may ended up with salable quantity 3, and -1 backordered. 

Getting back to how this functionality should be implemented conceptually, I would say that we no need to have this logic in 
\Magento\InventorySales\Plugin\StockState\CheckQuoteItemQtyPlugin which tracks  legacy \Magento\CatalogInventory\Api\StockStateInterface::checkQuoteItemQty  
instead we need to have this implemented as part of  \Magento\InventorySalesApi\Api\IsProductSalableForRequestedQtyInterface::execute  service 
Taking into account that this service returns \Magento\InventorySalesApi\Api\Data\ProductSalableResultInterface  
we can add backorder notification there as extension attribute of the ProductSalableResult. 

Also, from what I see we have a bug in Backorder Customer Notification functionality - \Magento\InventorySales\Model\IsProductSalableCondition\BackOrderNotifyCustomerCondition
    public function execute(string $sku, int $stockId, float $requestedQty): ProductSalableResultInterface
    {
        $stockItemConfiguration = $this->getStockItemConfiguration->execute($sku, $stockId);

        if ($stockItemConfiguration->getBackorders() === StockItemConfigurationInterface::BACKORDERS_YES_NOTIFY) {
            $stockItemData = $this->getStockItemData->execute($sku, $stockId);
            if (null === $stockItemData) {
                return $this->productSalableResultFactory->create(['errors' => []]);
            }

            $backOrderQty = $requestedQty - $stockItemData[GetStockItemDataInterface::QUANTITY];
which does not include quantity being reserved.

While Backorder functionality in general (letting to sell merchant more than he has in physical stock) should work. At least my assumption that conditions: 
\Magento\InventorySales\Model\IsProductSalableForRequestedQtyCondition\BackOrderCondition::execute
and 
\Magento\InventorySales\Model\IsProductSalableForRequestedQtyCondition\IsSalableWithReservationsCondition::execute
should handle this. 

just to make a summary.
I would make Backorder notification check being a part of `\Magento\InventorySalesApi\Api\IsProductSalableForRequestedQtyInterface::execute`, doing so we might be sure that it will always work as expected. 
If we have a limited timeframe - we can make a quick fix in BackOrderNotifyCustomerCondition , but I prefer that whole check to be moved under  IsProductSalableForRequestedQtyInterface 
  execution chain
Need to make sure whether case described in https://github.com/magento/inventory/issues/2454 is still reproducible and if so, what the root cause for it?

graham commented:
@gwharton -> I believe this is the correct graham :)

gwharton committed:
Merge branch '1.1-develop' into 2.3-develop

gwharton commented:
OK, I've completely hosed this PR trying to switch to 1.2-develop. Will be quicker to create new PR. Standby for updates to resolve review comments.

gwharton commented:
@maghamed As I was working on 1.1-develop branch, I failed to notice that the above PR had actually been picked up and merged into 1.2-develop in another process.

I agree with your comments above, and had already prepared updates, so I will create a new PR to incorporate your suggestion, but this can be approved in slow time as 1.2-develop actually DOES now have working backorders :)

maghamed commented:
sorry @graham, just copy-pasted my message from Slack. Unfortunately, the handles are not unique throughout :)

graham commented:
No worries. Have a great day!

On Tue, Jun 9, 2020, 13:21 Igor Miniailo <notifications@github.com> wrote:

> sorry @graham <https://github.com/graham>, just copy-pasted my message
> from Slack. Unfortunately, the handles are not unique throughout :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/magento/inventory/pull/2452#issuecomment-641548118>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAAKARIEV43457GY2BD4TTRV2KURANCNFSM4IJ3762Q>
> .
>

seansan commented:
Using 2.3.5 and are seeing same as OP. 

Created order with backorder qty = 2 (MSI stock = 0, allowed backorders -10)

qty_backordered is not populated in either quote (missing field) and in order items table where the field is always empty. Also we can not distinguish normal orders from backordered variants

Seeing the fix: will there also be a field to filter orders by? Resulting in a grid with orders that can be shipped immediately (without backorders) and orders that contain 1 or more backorders (and need to wait for the backorder to arrive)?

Also: what is the process with incoming stock uploaded to MSI? We assign the incoming stock to a source, then we can ship the order. There are no more processing orders: Then suppose the min_backorder_qty changs to -8 somewhere in time after that: is the result that the customer can order 8 again?

seansan commented:
In which release will we see this fix?

gwharton commented:
2.4

seansan commented:
> 2.4

thx really appreciate it (also amazed! this has not worked for so long)

Would be great to add to 2.3.6 too!!! We really need his code soon

seansan commented:
Is there some kind of workgroup for "backorders". We've been testing several use cases ... and together with MSI we see it is not a closed loop process. And it easily could be!! Where would one discuss this ... in a similar fashion as MSI ... let's call it workforce "Backorders" ;)

gwharton commented:
https://magentocommeng.slack.com/archives/C5FU5E2HY

seansan commented:
Thanks, sorry to put this here in git, bus seeing the greatest backorder enthusiasts are here. There is a M2 #backorder channel here https://magentocommeng.slack.com/archives/C016R7QHPEU

seansan commented:
Is there any way to pull in this code to 2.3.5? Or can we expect 2.3.6 soon - and the code will exist there too? (sorry but this all is very important for our process)

gwharton commented:
You will have to grab a diff from this PR and apply to 2.3.5 using composer patches. It applies just fine in 2.3.5. I am running it in production.

ShaunFM commented:
Can we pull this to 2.3.4? It would seem there is some new/added files in 2.3.5
